### PR TITLE
Ivanov Nikita. Lab 3. Option 1.

### PR DESCRIPTION
**Pull request by Atikin-NT**
Оптимизация, которая заменяет mul и add инструкцией muladd.

### CODE REVIEW

File: llvm/test/lab3/ivanov_nikita/muladdtest.mir [hunk 1]
```diff
@@ -0,0 +1,204 @@
...
+# }
+
+# __m128d test_two_ops(__m128d a, __m128d b, __m128d c, __m128d d) {
+#     __m128d tmp = a * b;
+#     a = a + b;
+#     b = b * b + a;
+#     d = tmp + c;
+#     return d * b;
+# }
+
```
Comment (m-ly4) [inline llvm/test/lab3/ivanov_nikita/muladdtest.mir:17]: Could you please add a test like:
```suggestion
# __m128d test_3(__m128d a, __m128d b, __m128d c, __m128d d) {
#     __m128d tmp = a * b;
#     __m128d tmp2 = tmp + c;
#     __m128d tmp3 = d + tmp;
#     return tmp2 / tmp3;
# }
```

**Review by m-ly4 — CHANGES_REQUESTED** [review]

**Comment by Atikin-NT** [comment]
@m-ly4, I have done all. You can check

File: llvm/lib/Target/X86/X86TargetMachine.cpp [hunk 2]
```diff
@@ -105,6 +105,7 @@ extern "C" LLVM_EXTERNAL_VISIBILITY void LLVMInitializeX86Target() {
   initializeX86ReturnThunksPass(PR);
   initializeX86DAGToDAGISelPass(PR);
   initializeX86ArgumentStackSlotPassPass(PR);
+  initializeAtikinPass(PR);
```
Comment (m-ly4) [inline llvm/lib/Target/X86/X86TargetMachine.cpp]: Please build a pass as dynamic lib to reduce impact on common files.

**Review by m-ly4 — CHANGES_REQUESTED** [review]

File: llvm/lib/Target/X86/X86TargetMachine.cpp (hunk 2 above)
Comment (Atikin-NT) [reply llvm/lib/Target/X86/X86TargetMachine.cpp]: How to do it? I tried to do it in a separate folder, but nothing worked. It just didn’t register my pass.

Comment (m-ly4) [reply llvm/lib/Target/X86/X86TargetMachine.cpp]: You may look into #100 as example.

Comment (Atikin-NT) [reply llvm/lib/Target/X86/X86TargetMachine.cpp]: ready

**Review by m-ly4 — APPROVED** [review]

**Pull request merged**